### ISSUE TITLE
HTTP: WebSocket Binary Messages Checks

### ISSUE THREAD

**Issue by xeizmendi** [gatling/gatling#2499]
At the moment there is no way to add checks to WebSocket binary messages, the messages are ignored.

**Comment by jbank** [comment]
Is anyone working on this? I would be happy to do the work but I won't start if someone is already working on it.

**Comment by slandelle** [comment]
@jbank Please go ahead, contributions welcome :)

**Comment by Lorkster** [comment]
Would be awesome to see support for binary messages as it would make testing for example Vertx and their websocket bridge much much easier!

**Comment by jbank** [comment]
I have implemented this based on how the ResponseBody is done. The only issue is I am not sure that web sockets are functioning since the upgrade to the latest AHC. Is anyone able to verify this for me?

**Comment by jbank** [comment]
I have done a bit more digging. Looks like the issue is the NettyWebSocket don't support using a LazyResponseBodyPartFactory as it tries to call getBodyPartBytes which is not supported by the LazyResponseBodyPart.

byte[] fragment = NettyResponseBodyPart.class.cast(part).getBodyPartBytes();

I tried to change the factory to the EagerResponseBodyPartFactory but gatling currently makes some assumptions that we are using the LazyResponseBodyPartFactory. Looks like we could change the code to call getBuf on the LazyResponseBodyPart. 

As I am very new to the code base it is hard for me to know which direction you would recommend.

**Comment by slandelle** [comment]
@jbank Thanks for reporting. I got this fixed in AHC, but I'm working on another issue atm. I'll push everything by the end of the week.

**Comment by jbank** [comment]
@slandelle Can you please provide an updated timeline for the AHC fix?

**Comment by slandelle** [comment]
@jbank done

**Comment by jbank** [comment]
@slandelle I haven't had a lot of time to debug this but it looks like if the within times out the socket gets into a bad state and never recovers. Hopefully I will have more time to debug this by the end of the week but I just wanted to see if it was on your radar?

**Comment by slandelle** [comment]
@jbank The AHC issue should be fixed in master. Did you give it a try?

**Comment by jbank** [comment]
Yea the AHC issue is fixed now. But when the websocket times out the websocket actor gets into a weird state.

**Comment by slandelle** [comment]
@jbank There's a chance you're referring to #2601. No, I really don't have the cycles for this atm.

**Comment by jbank** [comment]
Yea looks like it. Ok I will have a crack at it. I just didn't want to duplicate work.

**Comment by yuekang0629** [comment]
Hi, Is there any progress on this issue? Is there any way to check the binary message now?

**Comment by slandelle** [comment]
@jbank has done some initial work in #2882, looking forward to the update.

**Comment by jbank** [comment]
@yuekang0629 I will see if I can get to this today. Otherwise it will be in the next few weeks. This isn't my main priority. I am just doing it in spare cycles.

**Comment by slandelle** [comment]
Thanks @jbank !

**Comment by bharat-welingkar** [comment]
Checking in to see if you made any progress on this @jbank.

**Comment by fabienrenaud** [comment]
I see https://github.com/gatling/gatling/pull/2882 was rejected and that there won't be any real support for bytes until gatling 3, which isn't exactly around the corner...

How about at least unblocking folks who need to do byte testing by mapping the byte array to a string? https://github.com/gatling/gatling/blob/master/gatling-http/src/main/scala/io/gatling/http/action/async/ws/WsActor.scala?utf8=%E2%9C%93#L176 would then just call OnTextMessage which takes care of the rest.

**Comment by slandelle** [comment]
@fabienrenaud I understand how this would help for counting expected messages, but anything else is wrong:
* `equals` is not a valid check method, it's the plain old `equals` one
* binary payloads can't be arbitrarily converted to String (note you don't specify a Charset in the constructor, so you're using the default one for your JVM, hence platform dependent) as unmappable characters get replaced. People using binary frames do so because they use a binary encoding, typically protobuf, and you definitively expect to use the String bytes as the bytes -> String -> bytes would be destructive

In order to have the binary frames support move forward, we would need:
* a user who's really involved and provides real world use cases, test cases, and feedback
* time to implement

Ideally, this would be a sponsoring contract so we can dedicate some effort on this.

**Comment by fabienrenaud** [comment]
> equals is not a valid check method, it's the plain old equals one
Yeah I just found that out. regex and jsonpath checks are a bit limited.

> binary payloads can't be arbitrarily converted to String

Correct, unless the encoding of the binary payload to string is communicated and known by the tester. A binary payload converted to string could be matched against the expected binary payload converted with the same algorithm.
In my example, I just did a simple new String() on the byte array because i know i could match it exactly to the expected binary payload i would construct in a similar fashion in my scenario. Alternatively, the byte array could be base64 encoded and testers who'd like to use gatling with ws and bytes would know what to code against.

I tested this in the context of a protobuf app. It works and unblocks the usage of bytes in gatling2 for ws testing.

Again, I think this is better than just simply dropping messages on such an important feature and waiting 6-12 months for gatling 3 isn't an option for everyone.
